### ISSUE TITLE
npm postinstall hook runs before package-lock.json flushed to disk

### ISSUE THREAD

**Issue by cdaringe** [npm/npm#18798]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [x] Other (_see below for feature requests_):

#### What's going wrong?

- the `postinstall` hook runs before package-lock.json flushed to disk

i assume `postinstall` implies the install is complete, which includes the writing of the lockfile, however, this is not the case.  i need to update my registry origin URLs `postinstall` in the lockfile due to private registry malarkey.  this hook seems to be the only hook to enable that.

#### How can the CLI team reproduce the problem?

- register a postinstall hook to meddle w/ the lockfile
- install something
- observe your hook meddling fail

### supporting information:

 - `npm -v` prints: 5.4.2
 - `node -v` prints: 8.6.0
 - Windows, OS X/macOS, or Linux?: osx

**Comment by cdaringe** [comment]
i see i can use `postshrinkwrap`.  maybe a new task, `postlockfile`, would be helpful/more-clear

**Comment by jbruni** [comment]
Indeed, the `postinstall` hook should run after `package-lock.json` flushed to disk.
